clippy_lints/src/implicit_saturating_sub.rs

```diff
@@ -50,73 +52,229 @@ impl<'tcx> LateLintPass<'tcx> for ImplicitSaturatingSub {
+                condition.span,
+                expr.span,
+                right_hand,
+                left_hand,
+                if_block,
+                else_block,
+            ),
+            _ => {},
+        }
+    }
+}
 
-            // Check if assign operation is done
-            && let Some(target) = subtracts_one(cx, then)
+fn check_gt(
+    cx: &LateContext<'_>,
+    condition_span: Span,
+    expr_span: Span,
+    big_var: &Expr<'_>,
+    little_var: &Expr<'_>,
+    if_block: &Expr<'_>,
+    else_block: &Expr<'_>,
+) {
+    if let Some(big_var) = Var::new(big_var)
+        && let Some(little_var) = Var::new(little_var)
+    {
+        check_subtraction(cx, condition_span, expr_span, big_var, little_var, if_block, else_block);
+    }
+}
 
-            // Extracting out the variable name
-            && let ExprKind::Path(QPath::Resolved(_, ares_path)) = target.kind
+struct Var {
+    span: Span,
+    hir_id: HirId,
+}
+
+impl Var {
+    fn new(expr: &Expr<'_>) -> Option<Self> {
+        path_to_local(expr).map(|hir_id| Self {
+            span: expr.span,
+            hir_id,
+        })
+    }
+}
+
+fn check_subtraction(
+    cx: &LateContext<'_>,
+    condition_span: Span,
+    expr_span: Span,
+    big_var: Var,
+    little_var: Var,
+    if_block: &Expr<'_>,
+    else_block: &Expr<'_>,
+) {
+    let if_block = peel_blocks(if_block);
+    let else_block = peel_blocks(else_block);
+    if is_integer_literal(if_block, 0) {
+        // We need to check this case as well to prevent infinite recursion.
+        if is_integer_literal(else_block, 0) {
+            // Well, seems weird but who knows?
+            return;
+        }
+        // If the subtraction is done in the `else` block, then we need to also revert the two
+        // variables as it means that the check was reverted too.
+        check_subtraction(cx, condition_span, expr_span, little_var, big_var, else_block, if_block);
+        return;
+    }
+    if is_integer_literal(peel_blocks(else_block), 0)
+        && let ExprKind::Binary(op, left, right) = if_block.kind
+        && let BinOpKind::Sub = op.node
+    {
+        let local_left = path_to_local(left);
+        let local_right = path_to_local(right);
+        if Some(big_var.hir_id) == local_left && Some(little_var.hir_id) == local_right {
+            // This part of the condition is voluntarily split from the one before to ensure that
+            // if `snippet_opt` fails, it won't try the next conditions.
+            if let Some(big_var_snippet) = snippet_opt(cx, big_var.span)
+                && let Some(little_var_snippet) = snippet_opt(cx, little_var.span)
+            {
+                span_lint_and_sugg(
+                    cx,
+                    IMPLICIT_SATURATING_SUB,
+                    expr_span,
+                    "manual arithmetic check found",
+                    "replace it with",
+                    format!("{big_var_snippet}.saturating_sub({little_var_snippet})"),
```

Comment:
Well, if the compilation fails, the lint won't be run. So is it faster to check that it's not a const context and that MSRV > 1.46 or to run the check? :3

Reply:
Actually the MSRV is only for the const context. I'm not sure it's really useful here to check the const context for the reasons mentioned above.

Reply:
To give an example for what I mean (should've just done that directly, sorry), if I have a lib crate with an MSRV of 1.46 with code like
```rs
const fn stuff(a: u32, b: u32) {
  let c = if a > b { a - b } else { 0 };
}
```
then it wouldn't be useful if clippy told me that this is a manual implementation of `saturating_sub` and that I should use that instead. It was likely intentionally written that way because the method cant be used in const in that version, so would be a FP.

Can you explain this this part?
> So is it faster to check that it's not a const context and that MSRV > 1.46 or to run the check?"

What's the check?

Reply:
What I meant is that it won't compile if the compiler version used is invalid, so checking the MSRV will be useless in this case since clippy won't be run either.

And the check I was referring to was "is it run in a const context or not?".

Reply:
The code above compiles just fine in 1.46 though: https://godbolt.org/z/oT6EjvrWx, so I'd expect lints are run here.
So if clippy is suggesting `saturating_sub` there, then that won't work.

> So is it faster to check that it's not a const context and that MSRV > 1.46 or to run the check? :3

If I understand you correctly here now, we only need to avoid linting in const contexts if the msrv is set since nowadays it's const-stable, so we should have both checks, sth like `!(in_constant(expr) && !msrv.meets(SATURATING_SUB_CONST))`

Reply:
I don't why I was thinking that were linting on `saturating_sub` and not suggesting to add it... Tired brain, sorry for that. I'll add the const check.